LabDrivers/SIM900.py

```diff
@@ -24,99 +25,563 @@
+    properties['SIM980'] = {
+        'type':'multi',
+        'range':create_SIM980_range(),
+        'properties':create_SIM980_properties_obj()
+    }
+
+
+READ_BITS = 64
+NAME = 'SIM900'
 INTERFACE = Tool.INTF_SERIAL
 
 
 class Instrument(Tool.MeasInstr):
 
     def __init__(self, resource_name, debug=False, **kwargs):
-        super(Instrument, self).__init__(resource_name, 'SIM900', debug, interface=INTERFACE, baud_rate=9600,
-                                         term_chars="\n", timeout=2, bytesize=8, parity='N', stopbits=1, xonxoff=False, dsrdtr=False, **kwargs)
-
-    def ask_channel(self, chan_num, msg):
-        #        print 'SEND %i,"%s"'%(chan_num,msg)
-        self.write('CONN %i,"xyz"' % chan_num)
-        answer = self.ask('%s' % (msg))
-        i.write("xyz")
+        super(Instrument, self).__init__(resource_name, NAME, debug, interface=INTERFACE, baud_rate=9600,
+                                         term_chars="\n".encode(), timeout=2, bytesize=8, parity='N', stopbits=1, xonxoff=False, dsrdtr=False, **kwargs)
+        self.connections = {}
+        device_temp = {}
+        self.ports = ['1','2','3','4','5','6','7','8']#,'A','B','C','D']
+        print("Identifying devices")
+        self.connected_devices = {}
+        for port in self.ports:
+            resp = self.ask_channel(port,"*IDN?")
+            if resp == '' or resp == None:
+                print("There is no device at port "+port)
+            else:
+                self.connections[port] = resp.split(',')[1]
+                if self.connections[port] in self.connected_devices.keys():
+                    self.connected_devices[self.connections[port]].append(port)
+                else:
+                    self.connected_devices[self.connections[port]] = [port]
+                print(self.connections[port]+" detected on port "+port)
+                device_temp[self.connections[port]] = True
+        self.devices = device_temp.keys()
+
+        try:
+            if 'SIM980' in self.devices:
+                #print(self.connected_devices)
+                self.SIM980_PORT = self.connected_devices['SIM980'][0] # will only take first one
+            else:
+                self.SIM980_PORT = None
+        except:
+            print("error setting 'SIM980' port")
+    def measure(self, channel):
+        if channel == 'Summing Offset Voltage':
+            if self.SIM980_PORT is not None:
+                answer = self.read_offset_voltage(self.SIM980_PORT, AVERAGING_TIME)
+                answer = float(answer)
+            else:
+                answer = None
+        else:
+            print("invalid measurement channel: "+channel)
+            return -1
+        self.last_measure[channel] = answer
         return answer
-#        return i.ask('SNDT %i,"%s"'%(chan_num,msg))
 
-    def get_voltage(self, chan_num):
-        return self.ask_channel(chan_num, "VOLT?")
+    def get(self):
+        ret = {}
+        if 'SIM910' in self.devices:
```

Comment:
It is better to create a variable `SIM910 = 'SIM910'`
If you misspell the variable, the code will return an error, if you write `'SIM919'` instead of `'SIM910'`, the error will be indirect and more difficult to spot. The advice I received from my c++ teacher during my first year was : "if you use a value more than one, then why don't you assign it to a variable?"